### DISCUSSION TITLE
Список всех операций из расчетных таблиц цикла во временной таблице

### DISCUSSION THREAD

**Discussion by juramebel** [totumonline/totum-ru-issues-and-discussions#31] [Как это сделать?] [answered]
Добрый день!
Начал писать небольшой блок управления операциями в цеху по производству мебели.

Сейчас существует таблица циклов "Техкарты":
![3](https://user-images.githubusercontent.com/19141501/141728987-fab72463-0b9e-40b4-b8b7-6af5b5ce1bcc.PNG)

Внутри этой таблицы расчетные техкарты с операциями и нормами на конкретные заказы:
![2](https://user-images.githubusercontent.com/19141501/141728770-ccb8abd8-372a-460c-8cea-34c216fc0f1f.PNG)

Необходимо сделать интерфейс в виде одной таблицы для операторов, в котором будут отображаться все операции из этих расчетных таблиц, статус которых в таблице циклов, к примеру, «в работе». Чтобы работник цеха видел только список допущенных операций и не был вынужден каждый раз проваливаться внутрь карт (расчетных таблиц).

Но при этом чтобы он мог изменить статус операции из своего интерфейса (например на выполнено), и при необходимости провалиться в саму техкарту...

Есть ли какая то возможность решить эту задачу с помощью автозаполнения (у меня не получается обращаться из внешней временной таблицы внутрь всех расчетных)?

upd.
Еще такой вопрос. Каким самым простым способом можно организовать присвоение операции статуса текущей? С учетом того, что они идут жестко последовательно. Чтобы вынести эту операцию ярлыком в таблицу циклов.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### [Видео по этому вопросу на youtube —>](https://youtu.be/-Ls48xQ1Gx4)

Итак погнали. Общий подход такой:

1. Внутри цикла надо в футере надо сделать техническое поле в которое собрать `selectRowList` данные из строчной части с операциями.

![Снимок экрана 2021-11-15 в 15 32 32](https://user-images.githubusercontent.com/55755565/141783668-00108b28-cd03-4f6c-90a5-36ce8494834f.png)

Код в `teh_in_footer`:

```

= : selectRowList(table: $#ntn; field: 'field_1'; field: 'field_2'; field: 'field_3')

```

Или как вариант `teh_in_footer_2`:

```

= : listReplace(list: $sel; action: "field_3" = $#nci)

sel: selectRowList(table: $#ntn; field: 'field_1'; field: 'field_2')

```

2. В таблице Циклов сделать поле Ярлык и взять данные их этого поля изнутри футера или просто поле данные и взять функц `select` изнутри цикла.

![Снимок экрана 2021-11-15 в 15 41 45](https://user-images.githubusercontent.com/55755565/141783992-8ba16d7c-e868-419c-82f7-01d67ea9c3ce.png)

Здесь поле `link` это ярлык изнутри цикла, а `select_field` селект (надо один из вариантов использовать):

```

= : select(table: 's1_in_cycle'; field: 'f_teh_in_footer')

```

Не забудте пересчитать циклы и строки в таблице циклов после добавления поля.

> Это все затем, что бы потом выборка данных по статусу **не перлась внутрь циклов** — это сильно облегчает нагрузку. Это хорошая практика для циклов в Тотум.

3. Делаем временную таблицу — и делаем `h_insert` и `insert` — автозаполнение во временных тоже работает так-как они с Табличной единицей пересчета.

![Снимок экрана 2021-11-15 в 15 47 47](https://user-images.githubusercontent.com/55755565/141784688-c1061da9-36ed-4fa1-8870-7f417e81d061.png)

+ Отключите этой таблице добавление и удаление!

+ В `h_insert` собирается так: берем `selectList` из таблицы Циклов по статусу из созданного на шаге 2 поля — получится список-списков — надо будет его выпрямить в один список при помощи `listTrain`:

```

= : listTrain(list: $list)

list: selectList(table: 's1_cycles'; field: 'select_field'; where: 'status' = 1)

```

+ `insert` тип данные, тк у нас там row оказывается. Далее в строке мы берем значение кодом из него, например в поле `cycle`:

```

= : #insert[field_3]

```

+ К каждой строчке операций надо пришить номер цикла — чтоб можно было за допинфой в этот цикл сходить — это можно сделать сразу в шаге 1 — например сделайте техническое поле с кодом `$#nci` и возьмите его в набор полей, мы это делали на шаге 1.


4. Теперь у нас в временной таблице есть операции с номерами циклов и мы можем подтянуть чего угодно из цикла и выполнить действие в цикле из этой временной, например статус цикла. Код в `status`:

```

= : select(table: 's1_cycles'; field: 'status'; where: 'id' = #cycle)

```


5. Делаем кнопку «закрыть работу» с действием выполняемым в цикле и пересчетом временной. Код действия в кнопке:


```

a1= : set(table: 's1_in_cycle'; cycle: #cycle; field: 'h_status' = 2; log: true)

a2=: reCalculate(table: $#ntn)

```

**Reply by juramebel** [reply]
Спасибо огромное! Все получилось!
Я там потом еще небольшой вопрос написал, может и его подскажите?)

**Reply by juramebel** [reply]
Вроде придумал, как сделать текущую операцию. Через вычисления поля в хэдере при изменении состояния. 
Работает.

**Reply by ttm-alexeypolunin** [reply]
Тут я не совсем понял, кому присваивается статус. Циклу? Те если предыдущие циклы закрыты текущий становится в «активный» статус, так?

**Reply by juramebel** [reply]
Нет. 
Внутри расчетной таблицы "Техкарта" есть записи - операции.
У них свои последовательные номера.
При изменении статуса на завершено, активируется код, прибавляющий цифру к полю "Текущая операция" в хэдере.

**Reply by juramebel** [reply]
Но правда уже хочу перепридумать этот механизм.
Есть некоторые техпроцессы, где часть операций могут меняться местами в зависимости от ситуации в цеху. ПОжтому текущих операций может быть две

**Reply by ttm-alexeypolunin** [reply]
> Нет. Внутри расчетной таблицы "Техкарта" есть записи - операции. У них свои последовательные номера. При изменении статуса на завершено, активируется код, прибавляющий цифру к полю "Текущая операция" в хедере.

Еще одним путем является разместить это поле в футере и считать в нем количество выполненных операций + 1:

```

=: listSum(list: $list) + 1

list: selectList(table: $#ntn; field: 'id'; where: 'status' = "done")

```

Таким образом вам не нужно прорабатвать откат «выполнено» —> «в работе».

Поле нужно разместить в футере что бы обеспечить правильный порядок расчета. Его при этом можно отобразить в хедере используя параметр «отображать по другому плейсменту»

**Reply by juramebel** [reply]
Хм... отличный вариант)  спасибо!

**Comment by Xamius** [comment]
А как из этого примера вывести данные , только для определённого пользователя, который фильтруется по префильтру "пользователь" 
![image](https://user-images.githubusercontent.com/99695485/231447705-ac61add6-4d3d-474d-8785-35c157acf5e0.png)
И не видит другие циклы?
![image](https://user-images.githubusercontent.com/99695485/231447981-3fa99b39-5cc2-454a-b7d7-772551cfc069.png)
А только созданные им.
![image](https://user-images.githubusercontent.com/99695485/231448326-35671955-21d8-42f6-9155-7647e25b1fa0.png)

**Reply by Xamius** [reply]
Отвечу сам на свой вопрос:
```
= : listTrain(list: $list)

list: selectList(table: 'zayavka'; field: 'vyvod_dannyh_iz_ciklov'; where: 'polzovatel' = $if)
if: if(condition: $#nr = 2; then: $#nu; else: "*ALL*")

```

**Reply by ttm-alexeypolunin** [reply]
Угу, еще такой момент. Что бы этот пользователь не мог залезть внутрь чужих циклов — надо рестриктить ему это в настройках таблицы циклов — параметр называется Тип доступа к циклам или что-то типа того.

Тк Префильтр ограничивает ему доступ к текущей таблице, но не ограничивает доступ к таблицам внутри циклов.

**Reply by ttm-alexeypolunin** [reply]
И еще уточнение — если не требуется никакой калькуляции, то можно обойтись только ограничениями в Тип доступа к циклам. Если там поставить Видит только свои, он и в таблице Циклов не будет видеть чужих строк.

**Reply by Xamius** [reply]
А как указать несколько ролей в if: if(condition: $#nr = 2;
Что бы пользователи из группы 2 и 3 имели ограничение?

**Reply by erpic-ru** [reply]
if(condition: $#nr = json`[2,3]` then:... ?

**Reply by Xamius** [reply]
Да верно, 

> if(condition: $#nr = json`[2,3]` then:... ?

Спасибо

**Reply by Xamius** [reply]
А если настроены доступы к циклам "Начальник получает доступ создателя цикла к своим подчиненным" И есть допустим 2 начальника и у них есть по 3 подчинённых и они разбиты на роли. Как вывести именно циклы для начальников своих подчиненных?

**Reply by ttm-alexeypolunin** [reply]
Пользователь выбранный как Начальник для для другого пользователя автоматически видит строки циклов подчиненных.

Также если вы пользователю поставите Доступ к циклам для другого пользователя.

Так-как у вас там еще ограничения в префильтре для расчетов, то вам туда нужно взять всех пользователей, строки которых нужно показать. В таблице users для каждого пользователя есть калькулируемое поле Все пользователи доступа или типа того — можно использовать его.

**Reply by Xamius** [reply]
У меня есть вывод циклов во временную таблицу, и там выводиться вся информация по полям которые я указал 
``` = : selectRowList(table: $#ntn; field: 'field_1'; field: 'field_2'; field: 'field_3') ```
Мне надо сделать ограничение для начальников именно во временной таблице, что бы они в ней видели только своих подчиненных.
А префильтр и Доступ к циклам у меня настроены и работаю отлично
![image](https://user-images.githubusercontent.com/99695485/231783425-1ea2763c-f311-40ff-b91f-2aa88af83333.png)

А во временной таблицы начальники видят все заявки и они не группируются по подчиненным.

![image](https://user-images.githubusercontent.com/99695485/231784349-8e31f211-70e0-4997-ad3b-a321d8a6f188.png)

Хотя если применить код
```

 = : listTrain(list: $list)
list: selectList(table: 'zayavka'; field: 'vyvod_dannyh_iz_ciklov'; where: 'polzovatel' = $if)
if: if(condition: $#nr = 2; then: $#nu; else: "*ALL*") 
``` 
То подчиненные видят только свои заявки

![image](https://user-images.githubusercontent.com/99695485/231784517-027618d6-1d21-44f4-9982-11e7abe45846.png)

Вопрос в том , как вывести во временную таблицу для начальников, только тех кто им подчиняется

**Reply by ttm-alexeypolunin** [reply]
Поискать в таблице users тех, у кого они стоят как Начальник или взять оттуда же all_connected_users по начальнику и вычесть начальника — вы получите список id пользователей, к которым у начальника есть доступ.

**Reply by Xamius** [reply]
Отвечаю на свой вопрос 
```
= : listTrain(list: $list)

list: selectList(table: 'zayavka'; field: 'vyvod_dannyh_iz_ciklov'; where: 'polzovatel' = $if)
if: if(condition: $#nr = 1; then: "*ALL*"; else: $sel)

sel: select(table: 'users'; field: 'all_connected_users'; where: 'id' = $#nu)

```
